# 4.5 IC Data Consistency Checks

## Contents:

## Exercise

### 01. Importing Libraries

### 02. Importing Data

### 03. Mixed-Type Data

### 04. Missing Values

### 05. Duplicates

## Task

### 06. Mixed-Type Data

### 07. Missing Values

### 08. Duplicates

#### 01. Importing Libraries

In [289]:
# Immport libraries 
import pandas as pd
import numpy as np
import os

#### 02. Importing Data

In [290]:
# setting up path
path = r'/Users/yaseminmustafa/Desktop/CareerFoundry/Exercise 4/15-05-2025_Instacart Basket Analysis'

In [291]:
path

'/Users/yaseminmustafa/Desktop/CareerFoundry/Exercise 4/15-05-2025_Instacart Basket Analysis'

In [292]:
# Importing products dataframe
df_prods = pd.read_csv(os.path.join(path,"02_Data/Original Data/products.csv"), index_col = False)

In [293]:
# Importing wrangled orders
df_ords = pd.read_csv(os.path.join(path,"02_Data/Prepared Data/orders_wrangled.csv"), index_col = False)

#### 03. Mixed-Type Data

In [294]:
# Create a dataframe 
df_test = pd.DataFrame()

In [295]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [296]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [297]:
# Check for mixed types 
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [298]:
# Converting a columns data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')

#### 04. Missing Values

In [299]:
# Finding missing values inf products dataframe
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [300]:
# View these 16 values 
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [301]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14      12
68             69          NaN        26              7      12
115           116          NaN        93              3      11
261           262          NaN       110             13      12
525           525          NaN       109             11       1
1511         1511          NaN        84             16      14
1780         1780          NaN       126             11      12
2240         2240          NaN        52              1      14
2586         2586          NaN       104             13      12
3159         3159          NaN       126             11      13
3230         3230          NaN       120             16      14
3736         3736          NaN        41              8      15
4283         4283          NaN        77              7      14
4790         4790          NaN        91             16      14
38187       38183          NaN        39             12      21
40444       40440          NaN       120             16      15

In [302]:
df_prods.shape

(49693, 5)

#### 05. Duplicates

In [303]:
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [304]:
df_prods_clean.shape

(49677, 5)

In [305]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [306]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11       5  
18459        27              5       9  
26810        27              5      13  
35309       121             14       7  
35495       127             11      10

In [307]:
df_prods_clean.shape

(49677, 5)

In [308]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [309]:
df_prods_clean_no_dups.shape

(49672, 5)

### Task

In [310]:
df_ords.describe()

Unnamed: 0  order_id   user_id  order_number  orders_day_of_week  \
count   3,421,083 3,421,083 3,421,083     3,421,083           3,421,083   
mean    1,710,541 1,710,542   102,978            17                   3   
std       987,582   987,582    59,534            18                   2   
min             0         1         1             1                   0   
25%       855,270   855,272    51,394             5                   1   
50%     1,710,541 1,710,542   102,689            11                   3   
75%     2,565,812 2,565,812   154,385            23                   5   
max     3,421,082 3,421,083   206,209           100                   6   

       order_hour_of_day  days_since_prior_order  
count          3,421,083               3,214,874  
mean                  13                      11  
std                    4                       9  
min                    0                       0  
25%                   10                       4  
50%                   13                       7  
75%                   16                      15  
max                   23                      30

##### The days_since_prior_order column has fewer entries than the others, indicating missing values likely due to first-time orders, which have no prior reference point. Values range from 0 (same-day orders) to 30, suggesting a potential upper limit or monthly grouping. The order_number maxes out at 100, possibly indicating a cap per customer. order_dow ranges from 0 to 6, representing days of the week, and order_hour_of_day spans 0 to 23, covering a full day in hourly intervals. While order_id and user_id share the same count and minimum values, their differing maximums reflect that users can place multiple orders.

#### 06. Mixed-Type Data

In [311]:
# Step 3 - Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

#### 07. Missing Values

In [312]:
# Step 5 - Searching for missing values 
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
dataset_label                  0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

##### There are 206,209 missing values in the days_since_prior_order column. This is likely because these entries represent a customer's first-ever order in the dataset. Since there is no previous order to compare it to, it’s not possible to calculate the number of days since a prior order. As a result, it's expected that this field would be blank (NaN) for first-time orders.

In [313]:
df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [314]:
df_nan

Unnamed: 0  order_id  user_id dataset_label  order_number  \
0                 0   2539329        1         prior             1   
11               11   2168274        2         prior             1   
26               26   1374495        3         prior             1   
39               39   3343014        4         prior             1   
45               45   2717275        5         prior             1   
...             ...       ...      ...           ...           ...   
3420930     3420930    969311   206205         prior             1   
3420934     3420934   3189322   206206         prior             1   
3421002     3421002   2166133   206207         prior             1   
3421019     3421019   2227043   206208         prior             1   
3421069     3421069   3154581   206209         prior             1   

         orders_day_of_week  order_hour_of_day  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

In [315]:
df_ords.shape

(3421083, 8)

In [316]:
# Step 6 - Filtering the data that is not missing
df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [317]:
df_ords_clean.shape

(3214874, 8)

##### In this case, I chose not to impute or remove the missing values. Instead, I created a subset dataframe that contains only the rows where days_since_prior_order is not missing, and I continued my analysis using this new filtered dataset.

#### 08. Duplicates

In [318]:
# Step 7 - Checking for duplicates 
df_ords_dups = df_ords[df_ords.duplicated()]

In [319]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, dataset_label, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

##### no duplicates found.

In [320]:
# Step 9 - export df_ords_clean
prepared_data_path = os.path.join(path, '02_Data', 'Prepared Data')

In [321]:
os.makedirs(prepared_data_path, exist_ok = True) 

In [322]:
df_ords_clean.to_csv(os.path.join(prepared_data_path, 'orders_checked.csv'))

In [323]:
# Step 9 - export df_prods_clean_no_dups
df_prods_clean_no_dups.to_csv(os.path.join(prepared_data_path, 'products_checked.csv'))